File: src/Environment.cpp
#include "Idot/Environment.h"

#include "Idot/Diagnostics.h"

namespace idot {

Environment::Environment(std::shared_ptr<Environment> enclosing) : enclosing_(std::move(enclosing)) {}

void Environment::Define(const std::string& name, const Value& value) { values_[name] = value; }

void Environment::Assign(const Token& name, const Value& value) {
  auto found = values_.find(name.lexeme);
  if (found != values_.end()) {
    found->second = value;
    return;
  }

  if (enclosing_) {
    enclosing_->Assign(name, value);
    return;
  }

  throw DiagnosticError(ErrorPhase::Runtime, name.line, name.column,
                        "Undefined variable '" + name.lexeme + "'.");
}

Value Environment::Get(const Token& name) const {
  auto found = values_.find(name.lexeme);
  if (found != values_.end()) {
    return found->second;
  }

  if (enclosing_) {
    return enclosing_->Get(name);
  }

  throw DiagnosticError(ErrorPhase::Runtime, name.line, name.column,
                        "Undefined variable '" + name.lexeme + "'.");
}

}  // namespace idot

File: compiler/src/environment.rs
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

use crate::diagnostics::{DiagnosticError, ErrorPhase, Result};
use crate::token::Token;
use crate::value::Value;

#[derive(Debug, Clone)]
pub struct Environment {
    values: HashMap<String, Value>,
    enclosing: Option<Rc<RefCell<Environment>>>,
}

impl Environment {
    pub fn new(enclosing: Option<Rc<RefCell<Environment>>>) -> Self {
        Self {
            values: HashMap::new(),
            enclosing,
        }
    }

    pub fn define(&mut self, name: &str, value: Value) {
        self.values.insert(name.to_string(), value);
    }

    pub fn assign(&mut self, name: &Token, value: Value) -> Result<()> {
        if let Some(slot) = self.values.get_mut(&name.lexeme) {
            *slot = value;
            return Ok(());
        }

        if let Some(enclosing) = self.enclosing.as_ref().cloned() {
            return enclosing.borrow_mut().assign(name, value);
        }

        Err(DiagnosticError::new(
            ErrorPhase::Runtime,
            name.line,
            name.column,
            format!("Undefined variable '{}'.", name.lexeme),
        ))
    }

    pub fn get(&self, name: &Token) -> Result<Value> {
        if let Some(value) = self.values.get(&name.lexeme) {
            return Ok(value.clone());
        }

        if let Some(enclosing) = self.enclosing.as_ref().cloned() {
            return enclosing.borrow().get(name);
        }

        Err(DiagnosticError::new(
            ErrorPhase::Runtime,
            name.line,
            name.column,
            format!("Undefined variable '{}'.", name.lexeme),
        ))
    }
}
